tools/harness-simulation/harness/Thread_Harness/Sniffer/SimSniffer.py

```diff
@@ -0,0 +1,226 @@
+        return addr + ':' + port
+
+    @watched
+    def discoverSniffer(self):
+        sock = socket.socket(socket.AF_INET6, socket.SOCK_DGRAM)
+
+        # Define the output interface of the socket
+        ifn = self._find_index(IFNAME)
+        if ifn is None:
+            print('%s interface has not enabled IPv6.' % IFNAME)
+            return []
+        ifn = struct.pack('I', ifn)
+        sock.setsockopt(IPPROTO_IPV6, socket.IPV6_MULTICAST_IF, ifn)
+
+        # Send the request
+        sock.sendto(('sniffer').encode(), DISCOVERY_ADDR)
+
+        # Scan for responses
+        devs = set()
+        start = time.time()
+        while time.time() - start < SCAN_TIME:
+            if select.select([sock], [], [], 1)[0]:
+                addr, _ = sock.recvfrom(1024)
+                devs.add(addr)
+            else:
+                break
+
+        devs = [SimSniffer(addressofDevice=addr, channel=11) for addr in devs]
+        print(devs)
+
+        return devs
+
+    @watched
+    def startSniffer(self, channelToCapture, captureFileLocation, includeEthernet=False):
+        self.channel = channelToCapture
+        self._local_pcapng_location = captureFileLocation
+        self._remote_pcap_location = os.path.join(
+            REMOTE_SNIFFER_OUTPUT_PREFIX,
+            ''.join(random.choice(string.ascii_letters + string.digits) for _ in range(16)) + '.pcap'
+        )
+
+        self._ssh = paramiko.SSHClient()
+        self._ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+        remote_ip = self.ipaddr.split('@')[1]
+        self._ssh.connect(remote_ip, port=REMOTE_PORT, username=REMOTE_USERNAME, password=REMOTE_PASSWORD)
+        
+        _, stdout, _ = self._ssh.exec_command(
+            'echo $$ && exec python3 %s -o %s -c %d' % (
+                REMOTE_SNIFFER_PATH,
+                self._remote_pcap_location,
+                self.channel
+            )
+        )
+        self._remote_pid = int(stdout.readline())
+
+        self.log('local pcapng location = %s', self._local_pcapng_location)
+        self.log('remote pcap location = %s', self._remote_pcap_location)
+        self.log('remote pid = %d', self._remote_pid)
+
+        self.is_active = True
+
+    @watched
+    def stopSniffer(self):
+        if not self.is_active:
+            return
+        self.is_active = False
+
+        assert self._ssh is not None
+        self._ssh.exec_command('kill -s TERM %d' % self._remote_pid)
+        # wait to make sure the file is closed
+        time.sleep(3)
+        sftp = self._ssh.open_sftp()
+        # truncate suffix from .pcapng to .pcap
+        local_pcap_location = self._local_pcapng_location[:-2]
+        sftp.get(self._remote_pcap_location, local_pcap_location)
+        sftp.close()
+        self._ssh.close()
+
+        cmd = [EDITCAP_PATH, '-F', 'pcapng', local_pcap_location, self._local_pcapng_location]
+        self.log('running editcap: %r', cmd)
+        subprocess.Popen(cmd).wait()
```

Comment:
I think it is necessary to wait until `editcap` terminates. Because when the method returns, Harness will access and read the corresponding pcapng file. It will lead to unexpected consequences if `editcap` does not terminate.

Reply:
I meant that we need to eventually close the Popen object created by subprocess.Popen(cmd).

Reply:
Actually object `subprocess.Popen` does not have `close` or similar method. Referred to the [webpage](https://stackoverflow.com/a/32391463), the subprocess is gone after `wait()` from the view of the OS, the object `subprocess.Popen` does not own any system resources then.